File: quizzes/quizzesV5/quizzes9/17-problem.ts
export {};

const findSecondLargest = (args: number[]): number | null => {
  if (args.length < 2) return null;

  let largest: number = -Infinity;
  let secondLargest: number = -Infinity;

  for (const num of args) {
    if (num > largest) {
      [largest, secondLargest] = [num, largest];
    } else if (num > secondLargest && num < largest) {
      secondLargest = num;
    }
  }

  return secondLargest === -Infinity ? null : secondLargest;
};

console.log({ data: findSecondLargest([3, 1, 4, 1, 5, 9, 2, 6]) }); // Output: 6
console.log({ data: findSecondLargest([3]) }); // Output: null

File: bins/binsV2/bin/bins1/17-problem.rb
# frozen_string_literal: true

def find_second_largest(numbers)
  return { secondLargest: nil } if numbers.length < 2

  largest = -Float::INFINITY
  second_largest = -Float::INFINITY

  numbers.each do |num|
    if num > largest
      largest, second_largest = num, largest
    elsif num > second_largest && num != largest
      second_largest = num
    end
  end

  { secondLargest: second_largest === -Float::INFINITY ? nil : second_largest }
end

# Input
input = [ 3, 1, 4, 1, 5, 9, 2, 6 ]
puts find_second_largest(input)
puts find_second_largest([ 3 ])
# Output: {:secondLargest=>6}
